### ISSUE TITLE
KafkaConnect issue with SASL/Kerberos

### ISSUE THREAD

**Issue by chitenderkumar** [confluentinc/kafka-connect-storage-common#54]
trying to setup Confluent4.0.0 with SASL/Kerberos implemented. configured the Zookeeper, Kakfa, SchemaRegistry and Kafka-Rest with SASL/Kerberos. facing issue while executing connect.

below is the connect-avro-distributed.properties config
bootstrap.servers=SASL_PLAINTEXT://localhost:9092

group.id=connect-cluster

key.converter=io.confluent.connect.avro.AvroConverter
key.converter.schema.registry.url=http://localhost:8081
value.converter=io.confluent.connect.avro.AvroConverter
value.converter.schema.registry.url=http://localhost:8081

config.storage.topic=connect-configs
offset.storage.topic=connect-offsets
status.storage.topic=connect-statuses

config.storage.replication.factor=1
offset.storage.replication.factor=1
status.storage.replication.factor=1

internal.key.converter=org.apache.kafka.connect.json.JsonConverter
internal.value.converter=org.apache.kafka.connect.json.JsonConverter
internal.key.converter.schemas.enable=false
internal.value.converter.schemas.enable=false

plugin.path=share/java
security.protocol=SASL_PLAINTEXT
sasl.mechanism=GSSAPI
producer.sasl.mechanism=GSSAPI
producer.security.protocol=SASL_PLAINTEXT

consumer.sasl.mechanism=GSSAPI
consumer.security.protocol=SASL_PLAINTEXT

producer.confluent.monitoring.interceptor.sasl.mechanism=GSSAPI
producer.confluent.monitoring.interceptor.security.protocol=SASL_PLAINTEXT

consumer.confluent.monitoring.interceptor.sasl.mechanism=GSSAPI
consumer.confluent.monitoring.interceptor.security.protocol=SASL_PLAINTEXT

sasl.kerberos.service.name=kafka
sasl.jaas.config=[/opt/packages/confluent-4.0.0/etc/kafka/kafka_server_jaas.conf]

below is the connect Log,
'''[2018-01-04 10:03:42,553] INFO DistributedConfig values: 
	access.control.allow.methods = 
	access.control.allow.origin = 
	bootstrap.servers = [SASL_PLAINTEXT://localhost:9092]
	client.id = 
	config.storage.replication.factor = 1
	config.storage.topic = connect-configs
	connections.max.idle.ms = 540000
	group.id = connect-cluster
	heartbeat.interval.ms = 3000
	internal.key.converter = class org.apache.kafka.connect.json.JsonConverter
	internal.value.converter = class org.apache.kafka.connect.json.JsonConverter
	key.converter = class io.confluent.connect.avro.AvroConverter
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	offset.flush.interval.ms = 60000
	offset.flush.timeout.ms = 5000
	offset.storage.partitions = 25
	offset.storage.replication.factor = 1
	offset.storage.topic = connect-offsets
	plugin.path = [/opt/packages/confluent-4.0.0/share/java]
	rebalance.timeout.ms = 60000
	receive.buffer.bytes = 32768
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 40000
	rest.advertised.host.name = null
	rest.advertised.port = null
	rest.host.name = null
	rest.port = 8083
	retry.backoff.ms = 100
	sasl.jaas.config = [hidden]
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = kafka
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.mechanism = GSSAPI
	security.protocol = SASL_PLAINTEXT
	send.buffer.bytes = 131072
	session.timeout.ms = 10000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = null
	ssl.key.password = null
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.location = null
	ssl.keystore.password = null
	ssl.keystore.type = JKS
	ssl.protocol = TLS
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.location = null
	ssl.truststore.password = null
	ssl.truststore.type = JKS
	status.storage.partitions = 5
	status.storage.replication.factor = 1
	status.storage.topic = connect-statuses
	task.shutdown.graceful.timeout.ms = 5000
	value.converter = class io.confluent.connect.avro.AvroConverter
	worker.sync.timeout.ms = 3000
	worker.unsync.backoff.ms = 300000
 (org.apache.kafka.connect.runtime.distributed.DistributedConfig:223)
[2018-01-04 10:03:42,711] DEBUG Logging to org.slf4j.impl.Log4jLoggerAdapter(org.eclipse.jetty.util.log) via org.eclipse.jetty.util.log.Slf4jLog (org.eclipse.jetty.util.log:176)
[2018-01-04 10:03:42,713] INFO Logging initialized @35598ms (org.eclipse.jetty.util.log:186)
[2018-01-04 10:03:42,721] DEBUG org.eclipse.jetty.server.Server@62ee5b2a added {qtp789707827{STOPPED,8<=0<=200,i=0,q=0},AUTO} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,733] DEBUG HttpConnectionFactory@549ae373{HTTP/1.1} added {HttpConfiguration@4f349ae0{32768/8192,8192/8192,https://:0,[]},POJO} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,736] DEBUG ServerConnector@7f60e4{null}{0.0.0.0:0} added {org.eclipse.jetty.server.Server@62ee5b2a,UNMANAGED} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,736] DEBUG ServerConnector@7f60e4{null}{0.0.0.0:0} added {qtp789707827{STOPPED,8<=0<=200,i=0,q=0},AUTO} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,736] DEBUG ServerConnector@7f60e4{null}{0.0.0.0:0} added {org.eclipse.jetty.util.thread.ScheduledExecutorScheduler@55c07ec8,AUTO} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,737] DEBUG ServerConnector@7f60e4{null}{0.0.0.0:0} added {org.eclipse.jetty.io.ArrayByteBufferPool@56a2c988,POJO} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,737] DEBUG ServerConnector@7f60e4{null}{0.0.0.0:0} added {HttpConnectionFactory@549ae373{HTTP/1.1},AUTO} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,739] DEBUG ServerConnector@7f60e4{HTTP/1.1}{0.0.0.0:0} added {org.eclipse.jetty.server.ServerConnector$ServerConnectorManager@42f32b2,MANAGED} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,739] DEBUG org.eclipse.jetty.server.Server@62ee5b2a added {ServerConnector@7f60e4{HTTP/1.1}{0.0.0.0:8083},AUTO} (org.eclipse.jetty.util.component.ContainerLifeCycle:324)
[2018-01-04 10:03:42,908] DEBUG Registering Connect metrics with JMX for worker '10.0.4.93:8083' (org.apache.kafka.connect.runtime.ConnectMetrics:78)
[2018-01-04 10:03:42,909] INFO Kafka version : 1.0.0-cp1 (org.apache.kafka.common.utils.AppInfoParser:109)
[2018-01-04 10:03:42,909] INFO Kafka commitId : bedb2a8697fecd0d (org.apache.kafka.common.utils.AppInfoParser:110)
[2018-01-04 10:03:42,918] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;connector-startup-results (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,918] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;connector-startup-attempts (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,919] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;connector-startup-successes (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,919] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;connector-startup-failures (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,919] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;task-startup-results (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,920] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;task-startup-attempts (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,920] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;task-startup-successes (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,920] DEBUG Added sensor with name connect-sensor-group: connect-worker-metrics;task-startup-failures (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,940] DEBUG Added sensor with name connect-sensor-group: connect-worker-rebalance-metrics;completed-rebalance-count (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,940] DEBUG Added sensor with name connect-sensor-group: connect-worker-rebalance-metrics;rebalance-time (org.apache.kafka.common.metrics.Metrics:404)
[2018-01-04 10:03:42,947] DEBUG Updated cluster metadata version 1 to Cluster(id = null, nodes = [localhost:9092 (id: -1 rack: null)], partitions = []) (org.apache.kafka.clients.Metadata:270)
[2018-01-04 10:03:42,950] DEBUG [Worker clientId=connect-1, groupId=connect-cluster] The Connect group member has stopped. (org.apache.kafka.connect.runtime.distributed.WorkerGroupMember:177)'''

post this connect gets terminated.

any advice on this will be appreciated!

**Comment by rhauch** [comment]
It's hard to find the documentation (and we're already working to make it more clear), but the `sasl.jaas.config`, `producer.sasl.jaas.config`, and `consumer.sasl.jaas.config` properties all need to be specified (the `producer.*` and `consumer.*` properties do not inherit from the top-level worker properties; see [KAFKA-6387](https://issues.apache.org/jira/browse/KAFKA-6387)), and those properties do not take a path to a file. 

Instead, those properties must specify the JAAS login context parameters for SASL connections in the format used by JAAS configuration files. The JAAS configuration file format is described [here](http://docs.oracle.com/javase/8/docs/technotes/guides/security/jgss/tutorials/LoginConfigFile.html). The format for the value is: `<loginModuleClass> <controlFlag> (<optionName>=<optionValue>)*;`

**Comment by chitenderkumar** [comment]
@rhauch  Thank you for your response. we were frustrated with getting failed after and after with multiple configuration version. then we finally enabled debug log  on confluent and on kerberos server. and we found the issue with confluent service authenticating against each other through. for example earlier we had our kafka listener entry as `SASL_PLAINTEXT://localhost:9092` and kerberos principle as `kafka/HOST-FQDN@REALM-NAME`. with this conf Kafka was trying to authenticate itself for principle `kafka/localhost:9092` and it was failing to authenticate due to kerberos principle mismatch. 

post configuring the server.properties, zookeeper.properties and schema-registry.properties and connect-avro-distributed.properties with correct entries of bootstrap servers and the SASL/Kerberos properties confluent platform is running with all components( Zookeeper, Kafka, Schema-registry, Kafka-rest and Connect).
below are the entries of conf file:

**zookeeper.properties**
`dataDir=/tmp/zookeeper
clientPort=2181
maxClientCnxns=0
authProvider.1=org.apache.zookeeper.server.auth.SASLAuthenticationProvider
jaasLoginRenew=3600000
kerberos.removeHostFromPrincipal=true
kerberos.removeRealmFromPrincipal=true`

**server.properties**
`advertised.listeners=SASL_PLAINTEXT://10.0.4.93:9092
broker.id=0
confluent.support.customer.id=anonymous
confluent.support.metrics.enable=true
group.initial.rebalance.delay.ms=0
listeners=SASL_PLAINTEXT://10.0.4.93:9092
log.dirs=/tmp/kafka-logs
log.retention.check.interval.ms=300000
log.retention.hours=168
log.segment.bytes=1073741824
num.io.threads=8
num.network.threads=3
num.partitions=1
num.recovery.threads.per.data.dir=1
offsets.topic.replication.factor=1
sasl.enabled.mechanism=GSSAPI
sasl.kerberos.service.name=kafka
sasl.mechanism.inter.broker.protocol=GSSAPI
security.inter.broker.protocol=SASL_PLAINTEXT
socket.receive.buffer.bytes=102400
socket.request.max.bytes=104857600
socket.send.buffer.bytes=102400
transaction.state.log.min.isr=1
transaction.state.log.replication.factor=1
zookeeper.connection.timeout.ms=6000
zookeeper.connect=localhost:2181`

**schema-registry.properties**
`debug=false
kafkastore.bootstrap.servers=SASL_PLAINTEXT://10.0.4.93:9092
kafkastore.connection.url=localhost:2181
kafkastore.sasl.kerberos.service.name=kafka
kafkastore.sasl.mechanism=GSSAPI
kafkastore.security.protocol=SASL_PLAINTEXT
kafkastore.topic=_schemas
listeners=http://0.0.0.0:8081`

**connect-avro-distributed.properties**
`bootstrap.servers=SASL_PLAINTEXT://10.0.4.93:9092
config.storage.replication.factor=1
config.storage.topic=connect-configs
consumer.sasl.mechanism=GSSAPI
consumer.security.protocol=SASL_PLAINTEXT
group.id=connect-cluster
internal.key.converter=org.apache.kafka.connect.json.JsonConverter
internal.key.converter.schemas.enable=false
internal.value.converter=org.apache.kafka.connect.json.JsonConverter
internal.value.converter.schemas.enable=false
key.converter=io.confluent.connect.avro.AvroConverter
key.converter.schema.registry.url=http://localhost:8081
offset.storage.replication.factor=1
offset.storage.topic=connect-offsets
plugin.path=share/java
producer.sasl.mechanism=GSSAPI
producer.security.protocol=SASL_PLAINTEXT
sasl.kerberos.service.name=kafka
sasl.mechasim=GSSAPI
security.protocol=SASL_PLAINTEXT
status.storage.replication.factor=1
status.storage.topic=connect-statuses
value.converter=io.confluent.connect.avro.AvroConverter
value.converter.schema.registry.url=http://localhost:8081`
